subprojects/base-services/src/main/java/org/gradle/internal/reflect/DirectInstantiator.java

```diff
@@ -25,6 +25,23 @@
 import java.util.Arrays;
 
 public class DirectInstantiator implements Instantiator {
+    /*
+     * Every time we look up a constructor, we need to pass its types as a class array. Reusing the most common
+     * argument list sizes avoids a lot of garbage.
+     */
+    private static final ThreadLocal<Class[][]> REUSABLE_CLASS_ARRAYS = new ThreadLocal<Class[][]>() {
```

Comment:
I did some more measurements with all my changes and removing just this caching and there is no measureable difference in performance => Removed the change for simplicity.